src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs

```diff
@@ -0,0 +1,140 @@
+                    case { ChildOperationsCount: 0 }:
+                        return ImmutableArray<IOperation>.Empty;
+                    default:
+                        var builder = ArrayBuilder<IOperation>.GetInstance(Count);
+                        foreach (var child in this)
+                        {
+                            builder.Add(child);
+                        }
+                        return builder.ToImmutableAndFree();
+                }
+            }
+
+            IEnumerator<IOperation> IEnumerable<IOperation>.GetEnumerator() => this.GetEnumerator();
+            IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
+
+            public bool Any() => Count > 0;
+
+            public IOperation First()
+            {
+                var enumerator = GetEnumerator();
+                if (enumerator.MoveNext())
+                {
+                    return enumerator.Current;
+                }
+
+                throw new InvalidOperationException();
+            }
+
+            public Reversed Reverse() => new Reversed(_operation);
+
+            public IOperation Last()
+            {
+                var enumerator = Reverse().GetEnumerator();
+                if (enumerator.MoveNext())
+                {
+                    return enumerator.Current;
+                }
+
+                throw new InvalidOperationException();
+            }
+
+            /// <summary>
+            /// Implements a struct-based enumerator for <see cref="Operation"/> nodes. This type is not hardened
+            /// to <code>default(Enumerator)</code>, and will null reference in these cases. Calling <see cref="Current"/> after
+            /// <see cref="Enumerator.MoveNext"/> has returned false will throw an <see cref="InvalidOperationException"/>.
+            /// </summary>
+            [NonDefaultable]
+            public struct Enumerator : IEnumerator<IOperation>
+            {
+                /// <summary>
+                /// Implementation of the <see cref="Enumerator.MoveNext"/> and <see cref="Enumerator.Current"/>
+                /// members are delegated to the virtual <see cref="Operation.MoveNext(int, int)"/> and
+                /// <see cref="Operation.GetCurrent(int, int)"/> methods, respectively.
+                /// </summary>
+                private readonly Operation _operation;
+                /// <summary>
+                /// 
+                /// </summary>
+                private int _currentSlot;
+                private int _currentIndex;
+
+                internal Enumerator(Operation operation)
+                {
+                    _operation = operation;
+                    _currentSlot = -1;
+                    _currentIndex = -1;
+                }
+
+                public IOperation Current
+                {
+                    get
+                    {
+                        Debug.Assert(_operation != null && _currentSlot >= 0 && _currentIndex >= 0);
```

Comment:
>_currentSlot >= 0 && _currentIndex >= 0 [](http://example.com/codeflow?start=59&length=39)

`Current` is a `public` property so I don't think we can assert this. For instance, this will fail with `new Enumerator(operation).Current`.

Reply:
That is not a constructor you can call (publicly). The only way to obtain an instance of `Enumerator` is use `default`, or get one from an `Enumerable`. In either case, we should assert so that, in debug code, you know you're doing something wrong.